extensions/websocket.md

```diff
@@ -0,0 +1,38 @@
+---
+title: IRCv3 websocket support
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: Darren Whitlen
+    period: 2018
+    email: darren@kiwiirc.com
+---
+
+## Introduction
+
+Web based IRC clients are becoming more common and they all increase the demand for websocket connectivity to IRC networks.
+
+This specification provides a common implementation for web and other websocket clients to connect to IRC networks using common and standard development tools used on web pages, with a focus on a simple client facing interface.
+
+## WebSocket features and encoding
+
+Websockets offer different transports of communication, binary and strings. With the consideration that the IRC protocol is entirely lines of text, we make use of the message based text implementation that websockets natively provide. These are UTF8 encoded text messages where one websocket message consists of a single complete IRC message line.
+
+The internal websocket protocol handles message fragmentation and termination via frames, which is out of scope for this document.
+
+Because web browsers are pushing more towards a UTF8 default encoding for HTML5 documents and scriptings languages, and these web browsers are the primary use case for websockets, IRC servers MUST encode messages as UTF8 before sending them to the client. Ignoring this may result in the browser closing the websocket connection on invalid UTF8 sequences.
+
+## Message syntax
+
+The syntax of IRC lines MUST NOT be changed. However, IRC servers and clients MUST NOT include trailing new lines in their websocket messages as these are obsolete by websockets native messages.
```

Comment:
> IRC servers and clients MUST NOT include trailing new lines in their websocket messages 

I disagree with this, because it would conflict with RFC 2812 and RFC 1459.

Those RFCs state that IRC messages MUST end with `\r\n`, and that IRC is agnostic of the transport layer used (right now most use TCP, but WebSockets are allowed too).  This was made so IRC client libraries can be transport agnostic, and don't have to deal with special cases like messages over WebSockets not ending with `\r\n`.

Ideally WebSockets should just be another mean of transportation, and not an entirely new specification, as dwfreed said.  Clients and servers should threat WebSocket payloads as parts of a stream of data, like you'd do with TCP segments (not dealing with the text vs binary issue here).  This is what I think the RFCs had in mind anyway.

Reply:
Please note that I'm not against a "SHOULD" recommendation against putting several messages, or half a message in a WebSocket payload, but please let every IRC message be ended with their beloved `\r\n`.

Reply:
The new lines are used with TCP to mark the end of the line. Websockets have their own framing that renders that unnecessary and they’d just have to be stripped out before further processing anyway.

There’s already precedent for modifying the RFC message format in new extensions, I don’t see why CRLF is so special that it needs to be retained.

Reply:
>This was made so IRC client libraries can be transport agnostic, and don't have to deal with special cases like messages over WebSockets not ending with \r\n.

You don't need to deal with anything if you handle your line splitting separate from your message parsing and only implement it in your TCP transport layer.

Reply:
Hey!

> The new lines are used with TCP to mark the end of the line. 

That's where I disagree.  Newlines are used to mark the end of an *IRC message*.  Saying "*WebSockets already have their own framing, so when I implement IRC on top I don't have to put `\r\n` at the end*" is like saying "*my VPN-based layer 2 protocol have its own thing, so when I implement the Internet Protocol (layer 3) on top of it I don't have to put the destination address in the header*".

Maybe the example doesn't make much sense, but the point is to not mix the transport layer (TCP, WebSockets) with the application layer (IRC).

Reply:
> the point is to not mix the transport layer (TCP, WebSockets) with the application layer (IRC).

The RFCs mixed transport and application layers together as it was such a simple protocol back then. The application layer does not require `\r\n`, this is used as part of the transport to separate messages - framing. `\r\n` adds no meaning to a complete message.